### DISCUSSION TITLE
Document gotchas when user's invocation asks SEA to perform FS calls against own VFS

### DISCUSSION THREAD

**Discussion by cspotcode** [nodejs/single-executable#18] [General]
The [blog post](https://github.com/nodejs/single-executable/blob/main/blog/2022-08-05-an-overview-of-the-current-state.md) lists as one of the requirements of VFS:

> 7 | No interference with valid paths in the file system

Not the same as the above, but worth including in documentation for any VFS implementation: will those files be exposed via clever invocations of bundled executables?  This depends on where the VFS lives; what paths it uses.

For example if I bundle `prettier` into an executable and then run it, asking it to format a path that exists in the VFS: `./prettier ./prettier/__vfs__/index.js` then I'm guessing I will see bundled source emitted to stdout.

I think this is fine, it's just a characteristic of the VFS that may be worth documenting.  Wouldn't want users of SEA to assume bundled files are in any way secret without application-level safeguards.

---

Another thing that might crop up:

If I run `./globbing-tool './**/*'`, will it list the contents of the VFS?  If the VFS exists at `./globbing-tool` then `./globbing-tool/__vfs__/bin.js` *does* match the glob.

If I run `./globbing-tool './**/* | bash-stuff` then `bash-stuff` might try to read `./globbing-tool/__vfs__/bin.js` and fail.

**Comment by RaisinTen** [comment]
> will those files be exposed via clever invocations of bundled executables?

I don't think that's something we should expose to the end users because to them, it's just an executable file. If the single executable file is at location `./sea`, being able to access something like `./sea/__vfs__/index.js` is not something they should expect because that would treat `./sea` as a directory and that is incorrect behavior at least w.r.t how the OS allows you to access certain file paths.

Also, the files bundled into the single executable is an internal component of the bundled application, so I don't think there's any reason why an end user would want to access that. If it's something that an application author expects their users to access, IMO such a resource should be placed somewhere on the real file system.

**Reply by cspotcode** [reply]
That's my point: if you're saying that VFS is hidden, then you need to figure out how to hide it.  You need to decide where to put the VFS.  Where do you put it?

**Reply by ljharb** [reply]
The problem i see - which may be what @cspotcode is aiming at - is that from *outside*, the contents of the SEA would not be visible. But, from the *inside*, the SEA's code probably wants to be able to interact normally with BOTH filesystems - the user's, and the SEA.

Perhaps as a mounted disk, that's only available to code inside the SEA?

**Reply by cspotcode** [reply]
Yeah, exactly.  No matter how we design it, I expect there will limitations or gotchas to be documented.  For example, does code within the SPA need to use `require('vfs').readFileSync` to read bundled assets, because the normal FS module only reads the host filesystem?  Arguments can be made for or against; either way there will be limitations.

Does `require('/abs/path/to/gulp.config.js')` work to execute a user's config file within the SPA's VM?

If we enumerate these various scenarios in a spec, we can enumerate how they must be implemented/solved/documented for a given VFS implementation.

This list will give us a better idea of which modules can be bundled into an SPA as-is, and which use coding techniques that will need to be modified or special-cased to support the SPA runtime.

**Reply by ljharb** [reply]
I think that it would be *much* simpler if the VFS abstraction was invisible - ie, `fs` - and it simply magically "had access" to an additional mounted drive or path. This would make relative imports of SEA code, or relative pathnames to SEA files, Just Work, while still allowing access to the user's filesystem.

The user's code would not have access to this mount point, and so the internal SEA files would be encapsulated.

**Reply by cspotcode** [reply]
> Perhaps as a mounted disk, that's only available to code inside the SEA?

In the globbing example above, if code within the SPA does naive globbing -- using an off-the-shelf glob library -- then it might end up traversing the VFS.  This depends where the VFS is mounted and if `require('fs')` can see the VFS.  If `require('fs')` cannot see the VFS, then that causes other problems or limitations which can certainly be documented.

**Reply by ljharb** [reply]
I don't see that as a problem - the internal code can't be prevented from intentionally exposing internal files to the external environment, nor should it be.

**Reply by ovflowd** [reply]
Very interesting thread. I would also expect libraries that interact with the file system to be transparent to the VFS.

I honestly would imagine that in a standpoint of the original code that is now within the bundled SEA, i.e. my original call was

```
fs.readFileSync('./assets/file.json');
```

To be converted into

```
fs.readFileSync('__vfs__://./assets/file.json');
```

Or something of sort. I also would expect that once I execute the SEA bundler I would provide which paths should be included/mapped by SEA, respecting things like .seaignore (?)

So that if I intentionally had a relative file access that is not mapped there, eg.: SEA/Application has a local user config directory that needs to exist within the user file system, any "fs" call to that directory being a relative access or not, I would expect it to be respected by these options.

**Reply by ljharb** [reply]
That gets tricky tho when an fs operation is done dynamically, but i guess the vfs wrapper could do that dynamically as well.

**Reply by ovflowd** [reply]
Definitely! I would maybe go to lengths that maybe supporting dynamic imports could be tricky from day one. In the end, if the whole string gets constructed dynamically, we could instead of making the string interpolation within the fs statement, SEA could provide a Proxy for fs calls, and then analyze the final string that is sent to the fs methods

**Reply by cspotcode** [reply]
A list of scenarios can be created.  Then the list can be checked against any proposed VFS implementation, to see which items on the list are supported, which are not, which have workaround, which have limitations or gotchas worth documenting.  Depending on what goes on the list, not everything on the list needs to be supported.

A few items that can go on the list for starters: Static import, static require, dynamic import, dynamic require, passing `argv` values to FS calls, working directory is `/`, being able to manipulate VFS paths as strings vs `URL` objects

**Reply by jviotti** [reply]
I think we should definitely make the VFS transparent in the way that PKG (cc @jesec) has accomplished already. Otherwise, there are tons of existing modules out there that wouldn't work at all when packaged.

**Reply by RaisinTen** [reply]
FWIW, I have opened https://github.com/nodejs/single-executable/pull/28 to document all the requirements of the VFS, so feel free to put your suggestions there!
